### PR TITLE
Switched from nosetests to pytest

### LINKED ISSUE #661: Nosetests end-of-life
Nosetests does not work on Python 3.9 (and I was unable to get it working also on 3.8.6 on Windows). (See: https://tirkarthi.github.io/programming/2020/01/27/python-39-changes.html)

I think we shall switch to pytest like https://github.com/elastic/beats/pull/16883...

### PR DESCRIPTION
**Pull request by mikand**
Fixes #661

### CODE REVIEW

File: health_check.sh [hunk 1]
```diff
@@ -1,6 +1,4 @@
 #!/bin/bash
```
Comment (marcogario) [inline health_check.sh:1]: Let's drop this file. We were fetching more stats (e.g., coverage), but it seems we do not run it as part of CI. Easier to just drop it for now.

File: pysmt/solvers/z3.py [hunk 2]
```diff
@@ -147,14 +147,14 @@ def __init__(self, environment, logic, **options):
...
-        except OSError:
-            self.z3 = z3.Solver()
+        # try:
+        #     self.z3 = z3.SolverFor(str(logic))
+        # except z3.Z3Exception:
+        #     self.z3 = z3.Solver()
+        # except z3.z3types.Z3Exception:
+        #     self.z3 = z3.Solver()
+        # except OSError:
+        self.z3 = z3.Solver()
```
Comment (marcogario) [inline pysmt/solvers/z3.py:157]: This does not seem related to the pytest transition. Was this related to how we resolve the import of z3?

File: pysmt/test/smtlib/test_parser_qf_arrays.py [hunk 3]
```diff
@@ -16,10 +16,16 @@
...
+
 from pysmt.test.smtlib.parser_utils import execute_script_fname, SMTLIB_TEST_FILES, SMTLIB_DIR
 
-def test_generator():
+def _get_tests():
     for (logic, f, expected_result) in SMTLIB_TEST_FILES:
         smtfile = os.path.join(SMTLIB_DIR, f)
         if logic.theory.arrays:
-            yield execute_script_fname, smtfile, logic, expected_result
+            yield smtfile, logic, expected_result
```
Comment (marcogario) [inline pysmt/test/smtlib/test_parser_qf_arrays.py:27]: It seems we do not need a generator anymore. What about moving this into `pysmt.test` as :

```python
def smtlib_tests(logic_pred):
    tests = []
    for (logic, f, expected_result) in SMTLIB_TEST_FILES:
        smtfile = os.path.join(SMTLIB_DIR, f)
        if logic_pre(logic):
            tests.append(smtfile, logic, expected_result)
```

File: pytest.ini [hunk 4]
```diff
@@ -0,0 +1,3 @@
+[pytest]
+markers =
+    slow: marks tests as slow (deselect with '-m "not slow"')
```
Comment (marcogario) [inline pytest.ini:3]: Maybe add "(typically > 10 seconds)", to explain what we consider slow?

File: run_tests.sh [hunk 5]
```diff
@@ -17,13 +17,11 @@
...
-#export NOSE_PROCESSES=-1
-#export NOSE_PROCESS_TIMEOUT=240
 export PYTHONDONTWRITEBYTECODE=True
 
-# Skip slow tests (-A "not slow")
+# Skip slow tests (-m "not slow")
 # Exit on error (-x)
 # Rule of thumb: if a test takes more than 10 seconds it
 #                should be marked as slow using:
 #                    @attr("slow")
```
Comment (marcogario) [inline run_tests.sh:26]:
```suggestion
#                    @pytest.mark.slow
```

**Review by marcogario — APPROVED** [review]
LGTM.

There are some opportunities for refactoring the generators for the tests, that would be good to tackle now.

File: pysmt/solvers/z3.py (hunk 2 above)
Comment (mikand) [reply pysmt/solvers/z3.py:157]: I think this was the source of the damn segfault on the CI... But you are right, it is unrelated. Shall we move to a separate PR or keep it?

Comment (marcogario) [reply pysmt/solvers/z3.py:157]: Let's move it out and make sure this is indeed the reason of the CI segfault. I also do not remember why that code might have been there in the first place, so a more thorough review would be prefereable.

File: health_check.sh (hunk 1 above)
Comment (mikand) [reply health_check.sh]: OK

File: pysmt/solvers/z3.py (hunk 2 above)
Comment (mikand) [reply pysmt/solvers/z3.py]: OK

File: pysmt/test/smtlib/test_parser_qf_arrays.py (hunk 3 above)
Comment (mikand) [reply pysmt/test/smtlib/test_parser_qf_arrays.py]: Done.

File: run_tests.sh (hunk 5 above)
Comment (mikand) [reply run_tests.sh]: OK

**Pull request merged**